### PR TITLE
Dev cia 2

### LINKED ISSUE #5: develop unit tests

### PR DESCRIPTION
**Pull request by 4982**
Testmodule

### CODE REVIEW

**Comment by olivier-2018** [comment]
added merge issue parameters (reviewer, assignees, projects, linked issues, etc)

File: src_api/__init__.py [hunk 1]
```diff
@@ -14,8 +14,8 @@
 formatter = logging.Formatter("%(levelname)s:%(name)s:%(message)s")
 
 log_file = "app.log"
-if os.path.exists(log_file):
-    os.remove(log_file)
+#if os.path.exists(log_file):
+#    os.remove(log_file)
```
Comment (olivier-2018) [inline src_api/__init__.py]: should be uncommented.
the issue you experience is only related to windows file system access rights. 
It works fine on linux or deployed to Heroku.

File: tests/views_test.py [hunk 2]
```diff
@@ -1,5 +1,100 @@
-from src_api.views import helloworld
+from numpy import ndarray, result_type
+from src_api.views import *
+import shutil
+import os
 
+def TestBasics():
```
Comment (olivier-2018) [inline tests/views_test.py]: Capital letters are usually used for classes.
Please used best practice code formatting rules for functions (small caps)

File: tests/views_test.py [hunk 3]
```diff
@@ -1,5 +1,100 @@
...
+    Returns:
+        tuple: filenames and folder.
+    """
+    folder_name = os.path.join(app.config["APP_FOLDER"], "tests", "testfiles")
+    file_name = "test_file.wav"
+    file_name_copy = "test_file_copy.wav"
+    return folder_name,file_name,file_name_copy
+
+
+def TestBasicsCopy(shutil):
```
Comment (olivier-2018) [inline tests/views_test.py]: 1- See best practice formatting for functions (only use capital letters when declaring classes)
2- why do you use shutil as argument? it's imported above so available for all functions. No need to use it in the function call.

File: tests/views_test.py [hunk 4]
```diff
@@ -1,5 +1,100 @@
...
+    Returns:
+        tuple:  filenames and folder.
+    
+    """
+    folder_name, file_name, file_name_copy = TestBasics()
+    shutil.copyfile(os.path.join(folder_name,file_name), os.path.join(folder_name,file_name_copy))
+    return folder_name,file_name_copy
+
+
+def test_predict_from_file():
```
Comment (olivier-2018) [inline tests/views_test.py]: empty function, either implement test function or remove prior to merging with main branch

File: tests/views_test.py [hunk 5]
```diff
@@ -1,5 +1,100 @@
...
+    print()
+
+
+def test_backend_save_request_object():
```
Comment (olivier-2018) [inline tests/views_test.py]: empty function, either implement test function or remove prior to merging with main branch

**Review by 4982 — COMMENTED** [review]
ok

File: tests/views_test.py [hunk 6]
```diff
@@ -1,5 +1,100 @@
...
+        tuple:  filenames and folder.
+    
+    """
+    folder_name, file_name, file_name_copy = TestBasics()
+    shutil.copyfile(os.path.join(folder_name,file_name), os.path.join(folder_name,file_name_copy))
+    return folder_name,file_name_copy
+
+
+def test_predict_from_file():
+    print()
```
Comment (olivier-2018) [inline tests/views_test.py]: avoid using print() to skip code implementation.
use "pass" during development

**Comment by 4982** [comment]
will you merge?

File: tests/views_test.py [hunk 7]
```diff
@@ -1,5 +1,100 @@
...
+
+def test_backend_file_load():
+    print()
+
+
+def test_get_duration():
+    print()
+
+
+def test_get_sampling_rate():
```
Comment (olivier-2018) [inline tests/views_test.py:55]: i would suggest a few more test on this function using different files with different sampling rates to ensure the function returns the correct sampling rate each time.
I would suggest the use of pytest parametrize for this

File: tests/views_test.py [hunk 8]
```diff
@@ -1,5 +1,100 @@
...
+    """
+    Checks if the the Returned Sampling Rate of def_sampling_rate is 8000.
+    """
+    
+    folder_name, file_name_copy = TestBasicsCopy(shutil)
+    result = get_sampling_rate(file_name_copy,folder_name)
+    assert result == 8000
+
+
+def test_load_audio_sequence():
```
Comment (olivier-2018) [inline tests/views_test.py:66]: Same comment as above.
This function to return a sequence (numpy array) of float based on a specified sampling rate and max array dimension so one can use paramtrization to try multiple configurations and test if the output is always of the specified shape (not necessarily (8000,)

File: tests/views_test.py [hunk 9]
```diff
@@ -1,5 +1,100 @@
...
+
+def test_load_ML_model():
+    print()
+
+
+def test_make_prediction():
+    print()
+
+
+def test_backend_file_delete():
```
Comment (olivier-2018) [inline tests/views_test.py:80]: I would suggest to include making a copy of test files within the function.
WI don't think we can always guarantee that this function is always executed last.
Besides it does not have to be on the audio wav files, any created file (txt format, json, csv etc) should work with this delete function.
I would also test several different paths.

**Review by olivier-2018 — CHANGES_REQUESTED** [review]
please consider the comments from the review and resubmit a review.
Also, the test coverage is fairly low so we should try and add additional tests (complete the empty test functions) to cover more functionality for unit testing but also including integration testing, etc.
Try and make use of advanced pytest functionality we discussed in class.. parametrize, fixtures, etc

File: tests/views_test.py (hunk 3 above)
Comment (4982) [reply tests/views_test.py]: image

File: tests/views_test.py (hunk 4 above)
Comment (4982) [reply tests/views_test.py]: removed empty functions

File: tests/views_test.py (hunk 6 above)
Comment (4982) [reply tests/views_test.py]: deleted empty functions

File: tests/views_test.py (hunk 9 above)
Comment (4982) [reply tests/views_test.py:80]: delete is included in the functions

**Review by olivier-2018 — APPROVED** [review]
- app functionality checked.
- pytest check: all 4 tests passed
merge request can proceed..
Suggestions: add test cases to the parametrized test using different wav files, etc
